File: leetcode_learning/code_cpp/144.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//递归
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        PreOrder(root);
        return v;
    }
private:
    vector<int> v;
    void PreOrder(TreeNode* root)
    {
        if(root==NULL)
            return;
        v.push_back(root->val);
        PreOrder(root->left);
        PreOrder(root->right);
    }
};

//非递归
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> v;
        if(root==NULL)
            return v;
        stack<TreeNode*> s;
        while(root||!s.empty())
        {
            while(root)
            {
                v.push_back(root->val);
                s.push(root);
                root = root->left;
            }
            if(!s.empty())
            {
                root = s.top();
                s.pop();
                root = root->right;
            }
        }
        return v;
    }
};

File: leetcode_learning/auto-leetcode/144.二叉树的前序遍历/144.py
"""
144.二叉树的前序遍历
难度：简单
链接：https://leetcode.cn/problems/binary-tree-preorder-traversal/
"""
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # 递归解法：根 → 左 → 右
        result = []

        def traverse(node):
            if not node:
                return
            result.append(node.val)
            traverse(node.left)
            traverse(node.right)

        traverse(root)
        return result


class SolutionIterative:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # 迭代解法：用栈模拟，先右后左入栈（保证左子树先出栈）
        if not root:
            return []

        result = []
        stack = [root]

        while stack:
            node = stack.pop()
            result.append(node.val)
            # 先入右，后入左（栈后进先出，左子节点先被访问）
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)

        return result
